module-utils/utility/Utils.hpp

```diff
@@ -13,23 +13,109 @@
 #include <chrono>
 #include <random>
 #include <tuple>
+#include <type_traits>
 #include <functional>
 
 namespace utils
 {
     inline constexpr auto WHITESPACE = " \n\r\t\f\v";
-    constexpr unsigned int secondsInMinute =
-        std::chrono::duration_cast<std::chrono::seconds>(std::chrono::minutes(1)).count();
+    constexpr unsigned int secondsInMinute = 60;
 
-    std::string bytesToHex(const std::vector<std::uint8_t> &bytes);
-    std::vector<std::uint8_t> hexToBytes(const std::string &hex);
+    template <typename T> inline constexpr bool is_byte_v = std::is_integral_v<T> && sizeof(T) == sizeof(std::uint8_t);
 
-    template <typename T> std::string numToHex(T c)
+    template <typename T> using remove_cref_t = std::remove_const_t<std::remove_reference_t<T>>;
+
+    // NOTE: With short string optimization storing one byte hex in std::string is probably fine
+
+    template <typename T, std::enable_if_t<is_byte_v<T>, int> = 0> inline char halfByteToHex(T c)
+    {
+        return c < 10 ? '0' + c : 'a' + (c - 10);
+    }
+
+    // WARNING: Unsafe
+    template <typename T, std::enable_if_t<is_byte_v<T>, int> = 0>
+    inline void byteToHexAt(T c, std::string &result, std::size_t offset)
+    {
+        result[offset]     = halfByteToHex<std::uint8_t>((c & 0xF0) >> 4);
+        result[offset + 1] = halfByteToHex<std::uint8_t>(c & 0x0F);
+    }
+
+    template <typename T, std::enable_if_t<is_byte_v<T>, int> = 0> inline std::string byteToHex(T c)
+    {
+        std::string result(2, '0');
+        byteToHexAt(c, result, 0);
+        return result;
+    }
+
+    template <typename Bytes, std::enable_if_t<is_byte_v<typename Bytes::value_type>, int> = 0>
+    inline std::string bytesToHex(const Bytes &bytes)
+    {
+        std::string result(bytes.size() * 2, '0');
+        for (std::size_t i = 0; i < bytes.size(); ++i) {
+            byteToHexAt(bytes[i], result, i * 2);
+        }
+        return result;
+    }
+
+    template <typename T = std::uint8_t, std::enable_if_t<is_byte_v<T>, int> = 0> inline T halfHexToByte(char c)
+    {
+        if (c >= '0' && c <= '9') {
+            return c - '0';
+        }
+        else if (c >= 'A' && c <= 'F') {
+            return c - 'A' + 10;
+        }
+        else if (c >= 'a' && c <= 'f') {
+            return c - 'a' + 10;
+        }
+        throw std::invalid_argument("Unexpected hex digit");
+        return 0;
+    }
+
+    // WARNING: Unsafe
```

Comment:
could you please elaborate? 
Maybe there is a way to make it more safe?

Reply:
There is no check of the length of `hex` or value of `offset`. This is a function implementing common part of other functions and which should not be called directly.

Reply:
If I remove 2 of these functions then there is no need for these unsafe common parts to exist.

Reply:
Done